File: Src/Engine/Runtime/UI/GUI/Common/Image.cpp

#include "Image.h"

#include "Runtime/UI/GUI/Brushes/IBrush.h"

namespace SE
{
	Image::Image() : ContainerControl(0, 0, 64, 64), Brush()
	{
		AutoFocus = false;
	}

	Image::Image(float x, float y, float width, float height) : ContainerControl(x, y, width, height), Brush()
	{
		AutoFocus = false;
	}

	Image::Image(Float2 location, Float2 size) : ContainerControl(location, size), Brush()
	{
		AutoFocus = false;
	}

	Image::Image(Rectangle bounds) : ContainerControl(bounds), Brush()
	{
		AutoFocus = false;
	}

	void Image::DrawSelf()
	{
		ContainerControl::DrawSelf();

		if (Brush == nullptr)
			return;

		Rectangle rect;
		if (KeepAspectRatio)
		{
			// Figure out the ratio
			Float2 size = Size;
			Float2 imageSize = Brush->Size;
			if (imageSize.LengthSquared() < 1)
				return;
			Float2 ratio = size / imageSize;
			Float2 aspectRatio = ratio.MinValue();

			// Get the new height and width
			Float2 newSize = imageSize * aspectRatio;

			// Calculate the X,Y position of the upper-left corner
			// (one of these will always be zero)
			Float2 newPos = (size - newSize) / Float2(2);

			rect = Rectangle(newPos, newSize);
		}
		else
		{
			rect = Rectangle(Float2::Zero, Size);
		}

		Margin.ShrinkRectangle(rect);

		::SE::Color color = IsMouseOver || IsNavFocused ? MouseOverColor : Color;
		if (!Enabled)
			color *= DisabledTint;
		Brush->Draw(rect, color);
	}

	bool Image::OnMouseUp(Float2 location, MouseButton button)
	{
		if (ContainerControl::OnMouseUp(location, button))
			return true;

		if (Clicked.IsBinded())
		{
			Clicked(this, button);
			return true;
		}

		return false;
	}
} // SE

File: Src/Engine/Runtime/UI/GUI/Common/Image.cs
using System;

namespace SE.GUI
{
    /// <summary>
    /// Displays a managed brush and optionally reports clicks.
    /// </summary>
    public class Image : ContainerControl
    {
        private bool m_IsPressed;

        public Image()
        {
            AutoFocus = false;
        }

        public Image(Rectangle bounds)
            : this()
        {
            Bounds = bounds;
        }

        public event Action<Image, int>? Clicked;

        public IBrush? Brush { get; set; }
        public Margin Margin { get; set; }
        public Color Color { get; set; } = Color.White;
        public Color MouseOverColor { get; set; } = Color.White;
        public Color DisabledTint { get; set; } = new Color(128, 128, 128);
        public bool KeepAspectRatio { get; set; } = true;

        public override bool OnMouseDown(Float2 location, MouseButton button)
        {
            if (button != MouseButton.Left)
                return false;

            m_IsPressed = true;
            Root?.StartTrackingMouse(this);
            return true;
        }

        public override bool OnMouseUp(Float2 location, MouseButton button)
        {
            if (button != MouseButton.Left || !m_IsPressed)
                return false;

            m_IsPressed = false;
            Root?.EndTrackingMouse();
            if (location.X >= 0.0f && location.Y >= 0.0f && location.X <= Width && location.Y <= Height)
                Clicked?.Invoke(this, (int)button);
            return true;
        }

        public override void ClearState()
        {
            m_IsPressed = false;
            base.ClearState();
        }

        protected override void OnDraw()
        {
            base.OnDraw();
            if (Brush == null)
                return;

            Rectangle bounds = ScreenBounds;
            Margin.ShrinkRectangle(ref bounds);
            if (bounds.Width <= 0.0f || bounds.Height <= 0.0f)
                return;

            Color tint = Color;
            if (!EnabledInHierarchy)
                tint *= DisabledTint;
            else if (IsMouseOver)
                tint *= MouseOverColor;
            Brush.Draw(bounds, tint);
        }
    }
}
